cmd/prometheus/main.go

```diff
@@ -473,6 +473,16 @@ func main() {
 			cfg.tsdb.RetentionDuration = newFlagRetentionDuration
 		}
 
+		// When the retention time flag is set on config file it takes precedence.
+		if cfgFile.StorageConfig.TSDB.Retention.Time != 0 {
+			cfg.tsdb.RetentionDuration = cfgFile.StorageConfig.TSDB.Retention.Time
+		}
+
+		// When the retention size flag is set on config file it takes precedence.
+		if cfgFile.StorageConfig.TSDB.Retention.Size != 0 {
+			cfg.tsdb.MaxBytes = cfgFile.StorageConfig.TSDB.Retention.Size
+		}
+
 		if cfg.tsdb.RetentionDuration == 0 && cfg.tsdb.MaxBytes == 0 {
```

Comment:
There are few checks like this which needs to be done when we are re-applying the config
1. When both retention is set to 0
2. When retention duration is <0
3. (And any other checks that are here that I missed to mention).

Easiest way to do would be to create a wrapper around the ApplyConfig function for the localStorage

Example
```go
func wrapper(applyConfig) func(config) error {
	return func(config) error {
		// Do the validation
		// ...
		
		return applyConfig()
	}
}

```